ReactiveCocoa/Swift/Property.swift

```diff
@@ -8,66 +8,486 @@ public protocol PropertyType {
+	/// guarantee atomicity.
+	///
+	/// Returns the old value.
+	public func swap(newValue: Value) -> Value {
+		return modify { _ in newValue }
 	}
+}
+
+/// A read-only, type-erased view of a property.
+public struct AnyProperty<Value>: PropertyType {
+	private let box: AnyPropertyBoxBase<Value>
 
+	/// A producer for Signals that will send the wrapped property's current value,
+	/// followed by all changes over time, then complete when the wrapped property has
+	/// deinitialized.
 	public var producer: SignalProducer<Value, NoError> {
-		return _producer()
+		return box.producer
 	}
 
+	/// A signal that will send the wrapped property's changes over time, then complete
+	/// when the wrapped property has deinitialized.
 	public var signal: Signal<Value, NoError> {
-		return _signal()
+		return box.signal
 	}
-	
+
 	/// Initializes a property as a read-only view of the given property.
 	public init<P: PropertyType where P.Value == Value>(_ property: P) {
-		_value = { property.value }
-		_producer = { property.producer }
-		_signal = { property.signal }
+		box = AnyPropertyBox(property)
 	}
-	
+
 	/// Initializes a property that first takes on `initialValue`, then each value
 	/// sent on a signal created by `producer`.
+	///
+	/// The producer and the signal of the created property would complete only
+	/// when the `producer` completes.
 	public init(initialValue: Value, producer: SignalProducer<Value, NoError>) {
-		let mutableProperty = MutableProperty(initialValue)
-		mutableProperty <~ producer
-		self.init(mutableProperty)
+		self.init(propertyProducer: SignalProducer(value: initialValue).takeUntilReplacement(producer))
 	}
-	
+
 	/// Initializes a property that first takes on `initialValue`, then each value
 	/// sent on `signal`.
+	///
+	/// The producer and the signal of the created property would complete only
+	/// when the `signal` completes.
 	public init(initialValue: Value, signal: Signal<Value, NoError>) {
-		let mutableProperty = MutableProperty(initialValue)
-		mutableProperty <~ signal
-		self.init(mutableProperty)
+		self.init(propertyProducer: SignalProducer(value: initialValue).takeUntilReplacement(signal))
 	}
-}
 
-extension PropertyType {
-	/// Maps the current value and all subsequent values to a new value.
-	public func map<U>(transform: Value -> U) -> AnyProperty<U> {
-		let mappedProducer = SignalProducer<U, NoError> { observer, disposable in
-			disposable += ActionDisposable { self }
-			disposable += self.producer.map(transform).start(observer)
+	/// Initializes a property from a producer that promises to send at least one
+	/// value synchronously in its start handler before sending any subsequent event.
+	/// If the producer fails its promise, a fatal error would be raised.
+	///
+	/// The producer and the signal of the created property would complete only
+	/// when the `propertyProducer` completes.
+	private init(propertyProducer: SignalProducer<Value, NoError>, capturing closure: (() -> Void)? = nil) {
```

Comment:
Yeah, it would move the details from `lift` to `AnyProperty`. But the initialiser you commented on would still be there, as `init(initialValue:signal:)` and `init(initialValue:producer:)` now depends on it.